In [1]:
#1) Import necessary libraries

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [2]:
#2) Import dataset (exam_dataset.csv)

data = pd.read_csv(r"C:\Users\meorh\Desktop\Exam\exam_dataset.csv")
data.head()

Age     BusinessTravel  MonthlyIncome  JobSatisfaction  Bonus  \
0   41      Travel_Rarely           5993                4  17979   
1   49  Travel_Frequently           5130                2  20520   
2   37      Travel_Rarely           2090                3   6270   
3   33  Travel_Frequently           2909                3   8727   
4   27      Travel_Rarely           3468                2  10404   

               Department  DistanceFromHome  Education EducationField  \
0                   Sales                 1          2  Life Sciences   
1  Research & Development                 8          1  Life Sciences   
2  Research & Development                 2          2          Other   
3  Research & Development                 3          4  Life Sciences   
4  Research & Development                 2          1        Medical   

   EmployeeCount  ...                JobRole  MaritalStatus PerformanceRating  \
0              1  ...        Sales Executive         Single                 3   
1              1  ...     Research Scientist        Married                 4   
2              1  ...  Laboratory Technician         Single                 3   
3              1  ...     Research Scientist        Married                 3   
4              1  ...  Laboratory Technician        Married                 3   

   StockOptionLevel TrainingTimesLastYear WorkLifeBalance  YearsAtCompany  \
0                 0                     0               1               6   
1                 1                     3               3              10   
2                 0                     3               3               0   
3                 0                     3               3               8   
4                 1                     3               3               2   

   YearsSinceLastPromotion  OverTime  Attrition  
0                        0       Yes        Yes  
1                        1        No         No  
2                        0       Yes        Yes  
3                        3       Yes         No  
4                        2        No         No  

[5 rows x 24 columns]

In [3]:
data.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1470 entries, 0 to 1469
Data columns (total 24 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   Age                      1470 non-null   int64 
 1   BusinessTravel           1470 non-null   object
 2   MonthlyIncome            1470 non-null   int64 
 3   JobSatisfaction          1470 non-null   int64 
 4   Bonus                    1470 non-null   int64 
 5   Department               1470 non-null   object
 6   DistanceFromHome         1470 non-null   int64 
 7   Education                1470 non-null   int64 
 8   EducationField           1470 non-null   object
 9   EmployeeCount            1470 non-null   int64 
 10  EmployeeNumber           1470 non-null   int64 
 11  EnvironmentSatisfaction  1470 non-null   int64 
 12  Gender                   1470 non-null   object
 13  JobLevel                 1470 non-null   int64 
 14  JobRole                  1470 non-null  

In [4]:
#3) Allocate the relevant attributes as input and output

x = data.iloc[:,[0,1,2,3]].values
y = data.iloc[:,23].values

x,y

(array([[41, 'Travel_Rarely', 5993, 4],
        [49, 'Travel_Frequently', 5130, 2],
        [37, 'Travel_Rarely', 2090, 3],
        ...,
        [27, 'Travel_Rarely', 6142, 2],
        [49, 'Travel_Frequently', 5390, 2],
        [34, 'Travel_Rarely', 4404, 3]], dtype=object),
 array(['Yes', 'No', 'Yes', ..., 'No', 'No', 'No'], dtype=object))

In [5]:
#4) Use LabelEncoder to encode categorical data

from sklearn.preprocessing import LabelEncoder

labelencoder_x = LabelEncoder()
x[:,1] = labelencoder_x.fit_transform(x[:,1])

labelencoder_y = LabelEncoder()
y = labelencoder_y.fit_transform(y)

In [6]:
x,y

(array([[41, 2, 5993, 4],
        [49, 1, 5130, 2],
        [37, 2, 2090, 3],
        ...,
        [27, 2, 6142, 2],
        [49, 1, 5390, 2],
        [34, 2, 4404, 3]], dtype=object),
 array([1, 0, 1, ..., 0, 0, 0]))

In [7]:
#5) Split your data into training and test sets with the appropriate proportions

from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(x,y, test_size=0.2, random_state=0)

In [8]:
#6) Normalized your data using StandardScaler

from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
x_train = scaler.fit_transform(x_train)
x_test = scaler.transform(x_test)

In [9]:
#7) Fit the and predict results using the Classifier

from sklearn.ensemble import RandomForestClassifier

classifier = RandomForestClassifier(n_estimators=50, criterion='entropy', random_state=0)
classifier.fit(x_train, y_train)

prediction = classifier.predict(x_test)

In [10]:
y_test, prediction

(array([0, 0, 1, 0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0,
        0, 1, 0, 0, 1, 1, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0,
        1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1,
        1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 1,
        0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0,
        0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0,
        1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0,
        0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0,
        0, 1, 0, 0, 0, 1, 0, 0]),
 arr

In [11]:
#8) Evaluate your results using confusion matrix and calculate the prediction accuracy

from sklearn.metrics import confusion_matrix

cm = confusion_matrix(y_test, prediction)
print(cm)

[[233  12]
 [ 42   7]]


In [12]:
#8) Evaluate your results using confusion matrix and calculate the prediction accuracy

from sklearn.metrics import accuracy_score

accuracy = accuracy_score(y_test, prediction)
print(accuracy)

0.8163265306122449


In [13]:
#9) Discuss your results and findings

from sklearn.metrics import classification_report

report = classification_report(y_test, prediction)
print(report)


              precision    recall  f1-score   support

           0       0.85      0.95      0.90       245
           1       0.37      0.14      0.21        49

    accuracy                           0.82       294
   macro avg       0.61      0.55      0.55       294
weighted avg       0.77      0.82      0.78       294



Based on the accuracy value of 0.8163265306122449, column ‘Age’, ‘BusinessTravel’, ‘MonthlyIncome’ and ‘JobSatisfaction’ is quite good to measure column 'Attrition value'.